### DISCUSSION TITLE
Further fine-tuning Italian ASR model

### DISCUSSION THREAD

**Discussion by ftamburin** [NVIDIA-NeMo/NeMo#2508] [Q&A] [answered]
I am trying to further fine-tune using more data the Italian model STT_It_Quartznet15x5 (v.1.0.0rc1) provided by NVIDIA as described in the last Section of this documentation page

https://docs.nvidia.com/deeplearning/nemo/user-guide/docs/en/main/asr/configs.html

but I cannot find the config file used by NVIDIA for creating the Italian model.
Where can I find it?
In the folder "nemo/examples/asr/conf" I found some config for English and Chinese, but not that for Italian.
Could you help me?
Am I doing the right things?
I tried to extract it from the model .nemo file, but I was unsuccessful.
Thanks

**Comment by titu1994** [comment, marked as answer]
After loading the model, you can save the config using OmegaConf.save(model.cfg, filepath)

**Reply by ftamburin** [reply]
Thanks!

**Comment by ftamburin** [comment]
After configuring and loading the STT_it model correctly as suggested by @titu1994, I tried to further train the Italian model by using new speech data and the 'speech_to_text.py' script provided by NVIDIA, but without success.
From a starting WER around 0.15 (with the original Italian model) the training procedure jump to WER=0.37 at epoch 1 and then it does not improve at all. I tried also to unfreeze the encoder as suggested by one tutorial, but nothing happened.
The speech data are correctly encoded (16KHz, mono, 16bit) and the .json files correctly loaded and interpreted.
I am certainly making something wrong, but... 
Any idea? Could you help me?
Thanks

**Reply by titu1994** [reply]
Do you change the decoder vocabulary ? If not, your learning rate is possibly too high. Other case may be that the dataset is too small compared to MCV it dataset

**Reply by ftamburin** [reply]
I did not change the vocabulary because I started the new fine-tuning from the Italian model deployed from NVIDIA that already contains the right vocab. In the first experiment (the one described above) I used the same config contained in the NVIDIA model that contains a lr=0.0012 that looks smaller that the standard lr used when training from scratch (around 0.01?). In any case I made a second experiment decreasing the lr to 0.0005 and the results were similar: the initial WER jumps at epoch0 is smaller (0.25), but then the model does not improve anymore remaining more or less at the same WER on dev or even increase it. Could it be a problem of adjusting weight_decay?
The new dataset for training is more or less half of the original MCV one (102000 utterances vs 50000). Do you think this would be the problem?
Thanks!

**Reply by titu1994** [reply]
Use LR of 0.005 and train the model for longer. There is enough data (I think). The jump is expected as the model has to learn new words.

**Reply by ftamburin** [reply]
I followed your suggestion and trained the model with lr=0.005 for 100 epoch (with the data described before). 
After the starting jump in WER to 0.61, it started to decrease down to 0.40 with a reasonable loss reduction, but I am still very far from the starting WER before training (around 0.15). Should I train for more epoch? How many epoch should be enough in your opinion?
Many thanks again...

**Reply by okuchaiev** [reply]
@ftamburin another thing to check is that transcripts in your data don't use characters other than the ones in ``model.decoder.vocabulary`` 
For example, if you have numbers like "25" it should be converted to "twenty five", etc. Everything should be lowercased, punctuation removed, etc.

**Reply by ftamburin** [reply]
I followed your advices (@okuchaiev @titu1994 ) checking the training chars and restarting the training with lr=0.005, but, after 783000 iterations/233 epoch on a 32GB V100, the WER is stuck at 0.47 since epoch 70. 
I do not know how to improve the WER, now. Should I wait for more epoch? Are there some other problems in my training data? What do you think? 
Many thanks...

**Reply by titu1994** [reply]
This is odd, can you share your train script and config ? Perhaps any loss curve or metrics if available

**Reply by ftamburin** [reply]
This is the command I used for launching the training:

    python3 speech_to_text.py --config-path=../models --config-name=stt_it_quartznet15x5.yaml model.train_ds.manifest_filepath=TCorpora/ALLDATA.json model.validation_ds.manifest_filepath=cv-corpus-6.1-2020-12-11_dev.json trainer.max_epochs=1024 +init_from_nemo_model=../models/stt_it_quartznet15x5.nemo 

The script is the standard 'speech_to_text.py' from repository.
The loss profile during training:

![Screen Shot 2021-08-01 at 19 36 03](https://user-images.githubusercontent.com/15780295/127780427-4b7d75dd-0cc2-44a6-98ed-ad4a57df8070.png)

WER vs Epochs:
![Screen Shot 2021-08-01 at 19 45 59](https://user-images.githubusercontent.com/15780295/127780534-6710c70a-9a8b-479a-83d9-e3c15774cf7a.png)

And this is the yaml config file I used (mostly derived from the one embedded into stt_it_quartznet15x5 model):
[stt_it_quartznet15x5.txt](https://github.com/NVIDIA/NeMo/files/6913131/stt_it_quartznet15x5.txt)

Many thanks for your help

**Reply by ftamburin** [reply]
Hi,
any idea about the strange behavior of my training process?
I am making also an experiment starting from the STT_EN and fine-tuning for Italian adapting the tutorial 10_asr_ctc_language_finetuning.py, but I am experiencing the same behavior.  After some dozen of epochs the training stop improving... :(
Thanks

**Reply by titu1994** [reply]
If multiple models are showing similar results, then I would venture to guess that the issue is with the data rather than the models. Can you use the speech data explorer tool in Nemo to make sure that the transcripts make sense (ie there aren't empty transcripts or too many incorrect segmentations).

When following the notebook 10 for finetuning, do you try with Citrinet ? Could you build your own Tokenizer on Italian and fine-tune off one of the English Citrinet checkpoints to make sure it's not the data fault ? If Citrinet fails to train as well with a custom tokekizer, I would be quite sure the data is somehow incorrectly formatted. 

Finally, you can enable a flag in your "model" section of the config such that model.log_prediction=True from command line (or add the statement log_prediction: true inside the model section of yaml) so that during training the model will print out the ground truth and predicted text and visually ensure that there is no issue
